pkg/inject/inject.go

```diff
@@ -170,14 +182,17 @@ func (conf *ResourceConfig) GetPatch(
 		identity := k8s.TLSIdentity{
 			Name:                metaAccessor.GetName(),
 			Kind:                strings.ToLower(conf.meta.Kind),
-			Namespace:           "$" + podNamespaceEnvVarName,
+			Namespace:           "$" + envVarProxyPodNamespace,
 			ControllerNamespace: conf.globalConfig.GetLinkerdNamespace(),
 		}
 
 		report.update(conf)
 		if shouldInject(conf, report) {
 			conf.injectPodSpec(patch, identity)
 			conf.injectObjectMeta(patch)
+		} else {
+			log.Infof("skipped auto-inject on %s/%s", strings.ToLower(conf.meta.Kind), report.Name)
+			conf.overrideProxyConfigs(patch)
```

Comment:
It looks like this fixes the issue where a previously-injected deployment is re-injected when its annotations change. But I'm a bit nervous that this could have unintended consequences, especially since we're calling it on all resources that fail the `shouldInject` check, not just those with a linkerd sidecar. Plus the way in which we override containers is totally separate from the main inject code path. We might want to consider changing `shouldInject` to consider pods with existing linkerd proxies to be injectable, so that they go through the same code path, but I haven't given it enough thought. I hate to ask, but would it be possible to undo this part of this PR, and add support for honoring updated annotations in a separate PR? Even if we undo this change, this PR will still suffice for override annotations on create, which is a great start, and I just want to give the update use case a bit more thought, possibly in the larger context of #2339.

Reply:
Momentarily omitting this part sounds reasonable to me. At least per demo discussion, we are in agreement that the "override on updates" is a legit use case. I am hesitant to extend the `shouldInject` criteria, because IMO `shouldInject` and `overrideProxyConfigs` are two different tasks/responsibilities/features, with different resulting JSON patches. We'll chat more about it in the follow-up PR.

Reply:
Created #2498 to track this.